gcc/rust/ast/rust-ast-dump.cc

```diff
@@ -62,28 +57,86 @@ Dump::go (AST::Item &item)
   item.accept_vis (*this);
 }
 
+template <typename T>
 void
-Dump::format_function_param (FunctionParam &param)
+Dump::visit (T *node)
 {
-  param.get_pattern ()->accept_vis (*this);
-  stream << ": ";
-  param.get_type ()->accept_vis (*this);
+  node->accept_vis (*this);
 }
 
+template <typename T>
 void
-Dump::emit_attrib (const Attribute &attrib)
+Dump::visit (std::unique_ptr<T> &node)
 {
-  stream << "#[";
+  node->accept_vis (*this);
+}
+
+template <typename T>
+void
+Dump::visit_items_joined_by_separator (T &collection,
+				       const std::string &separator)
+{
+  if (collection.size () > 0)
+    {
+      visit (collection.at (0));
+      for (size_t i = 1; i < collection.size (); i++)
+	{
+	  stream << separator;
+	  visit (collection.at (i));
+	}
+    }
+}
+
+template <typename T>
+void
+Dump::visit_as_line (T &item, const std::string &trailing)
+{
+  stream << indentation;
+  visit (item);
+  stream << trailing << '\n';
+}
+
+template <typename T>
+void
+Dump::visit_items_as_lines (T &collection, const std::string &trailing)
+{
+  for (auto &item : collection)
+    visit_as_line (item);
+}
 
-  for (size_t i = 0; i < attrib.get_path ().get_segments ().size (); i++)
+template <typename T>
+void
+Dump::visit_items_as_block (T &collection, char left_brace, char right_brace)
+{
+  if (collection.empty ())
+    {
+      stream << left_brace << right_brace << '\n';
+    }
+  else
     {
-      const auto &seg = attrib.get_path ().get_segments ().at (i);
-      bool has_next = (i + 1) < attrib.get_path ().get_segments ().size ();
+      stream << left_brace << '\n';
 
-      stream << seg.get_segment_name ();
-      if (has_next)
-	stream << "::";
+      indentation.increment ();
+      visit_items_as_lines (collection, ",");
```

Comment:
I'm not sure I understand why block items should always have a trailing comma character. In the case of an extern block, that comma should be a semicolon if I'm understanding correctly.